src/FoxyClient.php

```diff
@@ -0,0 +1,637 @@
+use Psr\Http\Message\ResponseInterface;
+
+class FoxyClient
+{
+    const PRODUCTION_API_HOME = 'https://api.foxycart.com';
+    const SANDBOX_API_HOME = 'https://api-sandbox.foxycart.com';
+    const LINK_RELATIONSHIPS_BASE_URI = 'https://api.foxycart.com/rels/';
+    const PRODUCTION_AUTHORIZATION_ENDPOINT = 'https://my.foxycart.com/authorize';
+    const SANDBOX_AUTHORIZATION_ENDPOINT = 'https://my-sandbox.foxycart.com/authorize';
+    const DEFAULT_ACCEPT_CONTENT_TYPE = 'application/hal+json';
+    private static array $valid_config_options = [
+        'api_home',
+        'authorization_endpoint',
+        'use_sandbox',
+        'access_token',
+        'access_token_expires',
+        'refresh_token',
+        'client_id',
+        'client_secret',
+        'handle_exceptions'
+    ];
+    /**
+     * OAuth Access Token (can have various scopes such as client_full_access, user_full_access, store_full_access)
+     */
+    private string $access_token = '';
+    /**
+     * The timestamp for when the access token expires
+     */
+    private int $access_token_expires = 0;
+    /**
+     * OAuth Refresh token used to obtain new Access Token
+     */
+    private string $refresh_token = '';
+    /**
+     * OAuth endpoint for getting new Access Token
+     */
+    private ?string $oauth_token_endpoint = null;
+    /**
+     * Used by OAuth for getting new Access Token
+     */
+    private string $client_id = '';
+    /**
+     * Used by OAuth for getting new Access Token
+     */
+    private string $client_secret = '';
+    private ClientInterface $client;
+    private RequestFactoryInterface $requestFactory;
+    private ?ResponseInterface $last_response = null;
+    private array $registered_link_relations = ['self', 'first', 'prev', 'next', 'last'];
+    private array $links = [];
+    private bool $use_sandbox = false;
+    private bool $obtaining_updated_access_token = false;
+    private bool $include_auth_header = true;
+    private bool $handle_exceptions = true;
+    private string $accept_content_type = '';
+    private string $api_home = '';
+    private string $authorization_endpoint = '';
+
+    public function __construct(
+        RequestFactoryInterface $requestFactory,
+        ClientInterface $client,
+        array $config = []
+    ) {
+        $this->client = $client;
+        $this->requestFactory = $requestFactory;
+        $this->api_home = static::PRODUCTION_API_HOME;
+        $this->authorization_endpoint = static::PRODUCTION_AUTHORIZATION_ENDPOINT;
+        $this->updateFromConfig($config);
+    }
+
+    public function updateFromConfig(array $config): void
+    {
+        foreach (self::$valid_config_options as $valid_config_option) {
+            if (array_key_exists($valid_config_option, $config)) {
+                $this->{$valid_config_option} = $config[$valid_config_option];
+            }
+        }
+        if ($this->use_sandbox && (!array_key_exists('api_home', $config) || !array_key_exists(
```

Comment:
I guess this is the perfect use case for my comment I added earlier. If we define environment and store the URLs based on the environment then we can eliminate this redundant condition. 

@brettflorio